src/_pytest/pytester.py

```diff
@@ -1254,80 +1263,80 @@ def run(
         )
-        p1 = self.tmpdir.join("stdout")
-        p2 = self.tmpdir.join("stderr")
+        stdout_path = str(self.tmpdir.join("stdout"))
+        stderr_path = str(self.tmpdir.join("stderr"))
+
         print("running:", *cmdargs)
         print("     in:", py.path.local())
-        f1 = open(str(p1), "w", encoding="utf8")
-        f2 = open(str(p2), "w", encoding="utf8")
-        try:
-            now = timing.time()
+
+        stdout = open(stdout_path, "wb")
+        stderr = open(stderr_path, "wb")
+        with stdout, stderr:
+            start = timing.perf_counter()
+
             popen = self.popen(
                 cmdargs,
                 stdin=stdin,
-                stdout=f1,
-                stderr=f2,
+                stdout=stdout,
+                stderr=stderr,
                 close_fds=(sys.platform != "win32"),
             )
             if isinstance(stdin, bytes):
                 popen.stdin.close()
 
-            def handle_timeout() -> None:
-                __tracebackhide__ = True
-
-                timeout_message = (
+            try:
+                ret = popen.wait(timeout)
+            except subprocess.TimeoutExpired:
+                popen.kill()
+                popen.wait()
+                message = (
                     "{seconds} second timeout expired running:"
                     " {command}".format(seconds=timeout, command=cmdargs)
                 )
+                raise self.TimeoutExpired(message)
 
-                popen.kill()
-                popen.wait()
-                raise self.TimeoutExpired(timeout_message)
+            duration = timing.perf_counter() - start
+
+        stdout_text = open(stdout_path, encoding=encoding, errors=errors)
+        stderr_text = open(stderr_path, encoding=encoding, errors=errors)
+        with stdout_text, stderr_text:
+            out = stdout_text.read().splitlines()
+            err = stderr_text.read().splitlines()
 
-            if timeout is None:
-                ret = popen.wait()
-            else:
-                try:
-                    ret = popen.wait(timeout)
-                except subprocess.TimeoutExpired:
-                    handle_timeout()
-        finally:
-            f1.close()
-            f2.close()
-        f1 = open(str(p1), encoding="utf8")
-        f2 = open(str(p2), encoding="utf8")
         try:
-            out = f1.read().splitlines()
-            err = f2.read().splitlines()
-        finally:
-            f1.close()
-            f2.close()
-        self._dump_lines(out, sys.stdout)
-        self._dump_lines(err, sys.stderr)
+            for line in out:
+                print(line, file=sys.stdout)
+        except UnicodeEncodeError:
+            print("couldn't print to stdout because of encoding")
+        try:
+            for line in err:
+                print(line, file=sys.stderr)
+        except UnicodeEncodeError:
+            print("couldn't print to stderr because of encoding")
```

Comment:
these try/excepts would cause partial output to be written (and stopping at the problematic line) -- maybe consider writing everything at once instead so it's all-or-nothing

Reply:
A partial output is somewhat nicer for debug-ability I think?

My own preference would be to remove this except entirely and see if anyone complains :)